tools/oeedger8r/Emitter.ml

```diff
@@ -325,153 +380,139 @@ let oe_gen_args_header (ec : enclave_content) (dir : string) =
 
     NOTE: Foreign arrays are marshalled as [void *], but foreign pointers
     are marshalled as-is. *)
-let get_cast_to_mem_expr (ptype, decl) =
+let get_cast_to_mem_expr (ptype, decl) (parens : bool) =
   match ptype with
   | PTVal _ -> ""
   | PTPtr (t, _) ->
-      if is_array decl then sprintf "(%s*) " (get_tystr t)
+      let tystr = get_tystr t in
+      if is_array decl then
+        let s = tystr ^ "*" in
+        if parens then sprintf "(%s)" s else s
       else if is_foreign_array ptype then
-        sprintf "/* foreign array of type %s */ (void*) " (get_tystr t)
-      else sprintf "(%s) " (get_tystr t)
+        let s = if parens then "(void*)" else "void*" in
+        sprintf "/* foreign array of type %s */ %s" tystr s
+      else if parens then sprintf "(%s)" tystr
+      else tystr
 
-(** This function is nearly identical to the above, but does not
-    surround the expression with the final set of parentheses, which is
-    necessary for passing it to C macros. *)
-let get_cast_to_mem_type (ptype, decl) =
-  match ptype with
-  | PTVal _ -> ""
-  | PTPtr (t, _) ->
-      if is_array decl then get_tystr t ^ "*"
-      else if is_foreign_array ptype then
-        sprintf "/* foreign array of type %s */ void* " (get_tystr t)
-      else get_tystr t
+let oe_compute_input_buffer_size (plist : pdecl list) =
+  let params =
+    List.map
+      (fun (ptype, decl) ->
+        let size = oe_get_param_size (ptype, decl, "_args.") in
+        sprintf "if (%s) OE_ADD_SIZE(_input_buffer_size, %s);" decl.identifier
+          size )
+      (List.filter (fun p -> is_in_ptr p || is_inout_ptr p) plist)
+  in
+  (* Note that the indentation for the first line is applied by the
+     parent function. *)
+  if params <> [] then String.concat "\n    " params
+  else "/* There were no in nor in-out parameters. */"
+
+let oe_compute_output_buffer_size (plist : pdecl list) =
+  let params =
+    List.map
+      (fun (ptype, decl) ->
+        let size = oe_get_param_size (ptype, decl, "_args.") in
+        sprintf "if (%s) OE_ADD_SIZE(_output_buffer_size, %s);" decl.identifier
+          size )
+      (List.filter (fun p -> is_out_ptr p || is_inout_ptr p) plist)
+  in
+  (* Note that the indentation for the first line is applied by the
+     parent function. *)
+  if params <> [] then String.concat "\n    " params
+  else "/* There were no out nor in-out parameters. */"
+
+let oe_serialize_buffer_inputs (plist : pdecl list) =
```

Comment:
It feels like these duplication in these 3 functions can be avoided.
In future.

Reply:
I know, I felt that way while I was writing them, but just enough is different that it makes it difficult to deduplicate further. Their filter predicates are different (so their comments are different) and what they're printing is different. I opted for readability, but certainly it can be improved.